doc/specs/#888 - Com Api.md

```diff
@@ -0,0 +1,812 @@
+
+```c++ (C++ish pseudocode)
+m_installAppId = L"Microsoft.VSCode";
+```
+
+## 3.1. Create objects
+
+Creation of objects has to be done through CoCreateInstance rather than normal winrt initialization since it's hosted by an out of proc com server.
+These helper methods will be used in the rest of the examples.
+
+```c++ (C++ish pseudocode)
+    AppInstaller CreateAppInstaller()
+    {
+        com_ptr<abi::IAppInstaller> abiAppInstaller = nullptr;
+        check_hresult(CoCreateInstance(CLSID_AppInstaller, nullptr, CLSCTX_LOCAL_SERVER, IID_PPV_ARGS(abiAppInstaller.put())));
+        check_pointer(abiAppInstaller.get());
+        AppInstaller appInstaller{ abiAppInstaller.as<AppInstaller>() };
+        return appInstaller;
+    }
+    InstallOptions CreateInstallOptions()
+    {
+        com_ptr<abi::IInstallOptions> abiInstallOptions = nullptr;
+        check_hresult(CoCreateInstance(CLSID_InstallOptions, nullptr, CLSCTX_LOCAL_SERVER, IID_PPV_ARGS(abiInstallOptions.put())));
+        check_pointer(abiInstallOptions.get());
+        InstallOptions installOptions{ abiInstallOptions.as<winrt::Microsoft::Management::Deployment::InstallOptions>() };
+        return installOptions;
+    }
+    FindPackagesOptions CreateFindOptions()
+    {
+        com_ptr<abi::IFindPackagesOptions> abifindOptions = nullptr;
+        check_hresult(CoCreateInstance(CLSID_FindPackagesOptions, nullptr, CLSCTX_LOCAL_SERVER, IID_PPV_ARGS(abifindOptions.put())));
+        check_pointer(abifindOptions.get());
+        FindPackagesOptions findPackagesOptions{ abifindOptions.as<FindPackagesOptions>() };
+        return findPackagesOptions;
+    } 
+    GetCompositeAppCatalogOptions CreateGetCompositeAppCatalogOptions()
+    {
+        com_ptr<abi::IGetCompositeAppCatalogOptions> abiGetCompositeAppCatalogOptions = nullptr;
+        check_hresult(CoCreateInstance(CLSID_GetCompositeAppCatalogOptions), nullptr, CLSCTX_LOCAL_SERVER, IID_PPV_ARGS(abiGetCompositeAppCatalogOptions.put())));
+        check_pointer(abiGetCompositeAppCatalogOptions.get());
+        GetCompositeAppCatalogOptions getCompositeAppCatalogOptions{ abiGetCompositeAppCatalogOptions.as<GetCompositeAppCatalogOptions>() };
+        return getCompositeAppCatalogOptions;
+    }
+```
+
+## 3.2. Search
+
+The api can be used to search for packages in a catalog known to Windows Package Manager. This can be used to get availability information
+or start an install.
+
+```c++ (C++ish pseudocode)
+
+
+    IAsyncOperation<Package> FindPackageInCatalog(AppCatalog catalog, std::wstring packageId)
+    {
+        FindPackagesOptions findPackagesOptions = CreateFindOptions();
+        PackageMatchFilter filter;
+        filter.IsAdditive = true;
+        filter.Field = PackageMatchField::Id;
+        filter.Type = MatchType::Exact;
+        filter.Value = packageId;
+        findPackagesOptions.Filters().Append(filter);
+        FindPackagesResult findPackagesResult{ co_await catalog.FindPackagesAsync(findPackagesOptions) };

```

Comment:
We've been advocating for APIs to have a synchronous form unless the API is inherently async by doing network traffic and the like. Looks like that's what this API is doing.